File: app/src/main/java/com/opensource/gpstime/AnalogClockView.java
package com.opensource.gpstime;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.View;

import java.util.Calendar;
import java.util.Date;

public class AnalogClockView extends View {

    private static final float TICK_LONG_RATIO = 0.12f;
    private static final float TICK_SHORT_RATIO = 0.06f;

    private final Paint rimPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint tickPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint hourPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint minutePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint secondPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private final Paint centerPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    private int accentColor = 0xFF6750A4;
    private int faceColor = 0xFFFFFFFF;
    private int rimColor = 0xFFE0E0E0;
    private int tickColor = 0xFF49454F;

    private float hours = 0;
    private float minutes = 0;
    private float seconds = 0;

    public AnalogClockView(Context context) {
        super(context);
        init();
    }

    public AnalogClockView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    private void init() {
        tickPaint.setStyle(Paint.Style.STROKE);
        tickPaint.setStrokeCap(Paint.Cap.ROUND);

        hourPaint.setStyle(Paint.Style.STROKE);
        hourPaint.setStrokeCap(Paint.Cap.ROUND);

        minutePaint.setStyle(Paint.Style.STROKE);
        minutePaint.setStrokeCap(Paint.Cap.ROUND);

        secondPaint.setStyle(Paint.Style.STROKE);
        secondPaint.setStrokeCap(Paint.Cap.ROUND);

        centerPaint.setStyle(Paint.Style.FILL);

        TypedArray ta = getContext().getTheme().obtainStyledAttributes(new int[]{
                R.attr.clockAccentColor,
                R.attr.clockBackgroundColor,
                R.attr.clockTextColor,
                com.google.android.material.R.attr.colorOutline
        });
        accentColor = ta.getColor(0, 0xFF6750A4);
        faceColor = ta.getColor(1, 0xFFFFFFFF);
        tickColor = ta.getColor(2, 0xFF49454F);
        rimColor = ta.getColor(3, 0xFFE0E0E0);
        ta.recycle();
    }

    public void setAccentColor(int color) {
        accentColor = color;
        invalidate();
    }

    public void setFaceColor(int color) {
        faceColor = color;
        invalidate();
    }

    public void setRimColor(int color) {
        rimColor = color;
        invalidate();
    }

    public void setTime(Date date) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(date);
        hours = cal.get(Calendar.HOUR_OF_DAY) + cal.get(Calendar.MINUTE) / 60f;
        minutes = cal.get(Calendar.MINUTE) + cal.get(Calendar.SECOND) / 60f;
        seconds = cal.get(Calendar.SECOND) + cal.get(Calendar.MILLISECOND) / 1000f;
        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int size = Math.min(
                MeasureSpec.getSize(widthMeasureSpec),
                MeasureSpec.getSize(heightMeasureSpec)
        );
        setMeasuredDimension(size, size);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        float cx = getWidth() / 2f;
        float cy = getHeight() / 2f;
        float radius = Math.min(cx, cy) - 8;

        // Rim
        rimPaint.setColor(rimColor);
        rimPaint.setStyle(Paint.Style.STROKE);
        rimPaint.setStrokeWidth(4);
        canvas.drawCircle(cx, cy, radius, rimPaint);

        // Face
        rimPaint.setStyle(Paint.Style.FILL);
        rimPaint.setColor(faceColor);
        canvas.drawCircle(cx, cy, radius - 4, rimPaint);

        // Ticks
        tickPaint.setStrokeWidth(radius * 0.025f);
        for (int i = 0; i < 60; i++) {
            float angle = (float) Math.toRadians(i * 6 - 90);
            float tickLen = (i % 5 == 0) ? radius * TICK_LONG_RATIO : radius * TICK_SHORT_RATIO;
            if (i % 5 == 0) {
                tickPaint.setColor(accentColor);
                tickPaint.setStrokeWidth(radius * 0.04f);
            } else {
                tickPaint.setColor(tickColor);
                tickPaint.setStrokeWidth(radius * 0.02f);
            }
            float inner = radius - tickLen - 8;
            float x1 = cx + inner * (float) Math.cos(angle);
            float y1 = cy + inner * (float) Math.sin(angle);
            float x2 = cx + (radius - 8) * (float) Math.cos(angle);
            float y2 = cy + (radius - 8) * (float) Math.sin(angle);
            canvas.drawLine(x1, y1, x2, y2, tickPaint);
        }

        // Hour hand
        float hourAngle = (float) Math.toRadians(hours * 30 - 90);
        float hourLen = radius * 0.5f;
        hourPaint.setColor(accentColor);
        hourPaint.setStrokeWidth(radius * 0.08f);
        canvas.drawLine(cx, cy,
                cx + hourLen * (float) Math.cos(hourAngle),
                cy + hourLen * (float) Math.sin(hourAngle),
                hourPaint);

        // Minute hand
        float minAngle = (float) Math.toRadians(minutes * 6 - 90);
        float minLen = radius * 0.7f;
        minutePaint.setColor(accentColor);
        minutePaint.setStrokeWidth(radius * 0.05f);
        canvas.drawLine(cx, cy,
                cx + minLen * (float) Math.cos(minAngle),
                cy + minLen * (float) Math.sin(minAngle),
                minutePaint);

        // Second hand
        float secAngle = (float) Math.toRadians(seconds * 6 - 90);
        float secLen = radius * 0.75f;
        secondPaint.setColor(accentColor);
        secondPaint.setStrokeWidth(radius * 0.015f);
        canvas.drawLine(cx, cy,
                cx + secLen * (float) Math.cos(secAngle),
                cy + secLen * (float) Math.sin(secAngle),
                secondPaint);

        // Center dot
        centerPaint.setColor(accentColor);
        canvas.drawCircle(cx, cy, radius * 0.06f, centerPaint);
        centerPaint.setColor(accentColor);
        centerPaint.setAlpha(180);
        canvas.drawCircle(cx, cy, radius * 0.025f, centerPaint);
    }
}

File: app/src/main/java/com/opensource/gpstime/AnalogClockView.kt
package com.opensource.gpstime

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import android.view.View
import com.google.android.material.R
import java.util.Calendar
import java.util.Date

class AnalogClockView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null
) : View(context, attrs) {

    private val rimPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val tickPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val hourPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val minutePaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val secondPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val centerPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private var accentColor = 0xFF6750A4.toInt()
    private var faceColor = -0x1
    private var rimColor = 0xFFE0E0E0.toInt()
    private var tickColor = 0xFF49454F.toInt()

    private var hours = 0f
    private var minutes = 0f
    private var seconds = 0f

    init {
        tickPaint.style = Paint.Style.STROKE
        tickPaint.strokeCap = Paint.Cap.ROUND
        hourPaint.style = Paint.Style.STROKE
        hourPaint.strokeCap = Paint.Cap.ROUND
        minutePaint.style = Paint.Style.STROKE
        minutePaint.strokeCap = Paint.Cap.ROUND
        secondPaint.style = Paint.Style.STROKE
        secondPaint.strokeCap = Paint.Cap.ROUND
        centerPaint.style = Paint.Style.FILL

        val ta = context.theme.obtainStyledAttributes(intArrayOf(
            com.opensource.gpstime.R.attr.clockAccentColor,
            com.opensource.gpstime.R.attr.clockBackgroundColor,
            com.opensource.gpstime.R.attr.clockTextColor,
            R.attr.colorOutline
        ))
        accentColor = ta.getColor(0, 0xFF6750A4.toInt())
        faceColor = ta.getColor(1, -0x1)
        tickColor = ta.getColor(2, 0xFF49454F.toInt())
        rimColor = ta.getColor(3, 0xFFE0E0E0.toInt())
        ta.recycle()
    }

    fun setAccentColor(color: Int) {
        accentColor = color
        invalidate()
    }

    fun setFaceColor(color: Int) {
        faceColor = color
        invalidate()
    }

    fun setRimColor(color: Int) {
        rimColor = color
        invalidate()
    }

    fun setTime(date: Date) {
        val cal = Calendar.getInstance()
        cal.time = date
        hours = cal.get(Calendar.HOUR_OF_DAY) + cal.get(Calendar.MINUTE) / 60f
        minutes = cal.get(Calendar.MINUTE) + cal.get(Calendar.SECOND) / 60f
        seconds = cal.get(Calendar.SECOND) + cal.get(Calendar.MILLISECOND) / 1000f
        invalidate()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val size = Math.min(
            MeasureSpec.getSize(widthMeasureSpec),
            MeasureSpec.getSize(heightMeasureSpec)
        )
        setMeasuredDimension(size, size)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        val cx = width / 2f
        val cy = height / 2f
        val radius = Math.min(cx, cy) - 8f

        rimPaint.color = rimColor
        rimPaint.style = Paint.Style.STROKE
        rimPaint.strokeWidth = 4f
        canvas.drawCircle(cx, cy, radius, rimPaint)

        rimPaint.style = Paint.Style.FILL
        rimPaint.color = faceColor
        canvas.drawCircle(cx, cy, radius - 4f, rimPaint)

        tickPaint.strokeWidth = radius * 0.025f
        for (i in 0 until 60) {
            val angle = Math.toRadians((i * 6 - 90).toDouble())
            val tickLen = if (i % 5 == 0) radius * TICK_LONG_RATIO else radius * TICK_SHORT_RATIO
            if (i % 5 == 0) {
                tickPaint.color = accentColor
                tickPaint.strokeWidth = radius * 0.04f
            } else {
                tickPaint.color = tickColor
                tickPaint.strokeWidth = radius * 0.02f
            }
            val inner = radius - tickLen - 8f
            val x1 = cx + inner * Math.cos(angle).toFloat()
            val y1 = cy + inner * Math.sin(angle).toFloat()
            val x2 = cx + (radius - 8f) * Math.cos(angle).toFloat()
            val y2 = cy + (radius - 8f) * Math.sin(angle).toFloat()
            canvas.drawLine(x1, y1, x2, y2, tickPaint)
        }

        val hourAngle = Math.toRadians((hours * 30 - 90).toDouble())
        val hourLen = radius * 0.5f
        hourPaint.color = accentColor
        hourPaint.strokeWidth = radius * 0.08f
        canvas.drawLine(cx, cy,
            cx + hourLen * Math.cos(hourAngle).toFloat(),
            cy + hourLen * Math.sin(hourAngle).toFloat(),
            hourPaint)

        val minAngle = Math.toRadians((minutes * 6 - 90).toDouble())
        val minLen = radius * 0.7f
        minutePaint.color = accentColor
        minutePaint.strokeWidth = radius * 0.05f
        canvas.drawLine(cx, cy,
            cx + minLen * Math.cos(minAngle).toFloat(),
            cy + minLen * Math.sin(minAngle).toFloat(),
            minutePaint)

        val secAngle = Math.toRadians((seconds * 6 - 90).toDouble())
        val secLen = radius * 0.75f
        secondPaint.color = accentColor
        secondPaint.strokeWidth = radius * 0.015f
        canvas.drawLine(cx, cy,
            cx + secLen * Math.cos(secAngle).toFloat(),
            cy + secLen * Math.sin(secAngle).toFloat(),
            secondPaint)

        centerPaint.color = accentColor
        canvas.drawCircle(cx, cy, radius * 0.06f, centerPaint)
        centerPaint.color = accentColor
        centerPaint.alpha = 180
        canvas.drawCircle(cx, cy, radius * 0.025f, centerPaint)
    }

    companion object {
        private const val TICK_LONG_RATIO = 0.12f
        private const val TICK_SHORT_RATIO = 0.06f
    }
}
